crates/circuit/src/operations.rs

```diff
@@ -2087,6 +2088,27 @@ pub struct PyInstruction {
     pub instruction: PyObject,
 }
 
+impl PyInstruction {
+    pub fn blocks(&self) -> Option<impl Iterator<Item = CircuitData>> {
+        Python::with_gil(|py| -> Option<vec::IntoIter<CircuitData>> {
```

Comment:
If this needs the GIL, can we just accept the GIL token in the function call?  The PR comment says this is unlikely to be the final interface, and removing the GIL token from all the call sites won't be hard when it comes to it.

I'm concerned about how much we're hiding `with_gil` calls in Qiskit at the moment - given that the move to Rust is in part meant to make threaded parallelism easier for us, I think hiding implicit serialisation is possibly working against us.

Reply:
I moved `blocks` to the `Operation` trait, and it seems like that interface doesn't accept a `Python` token for any of the methods. I agree with your concern, but keeping it out of the signature feels more consistent with everything else, for now.